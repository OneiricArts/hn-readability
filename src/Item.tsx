import React, { useEffect, useState, useRef, RefObject } from 'react';
import ReactDOM from 'react-dom';
import DOMPurify from 'dompurify';
import { Collapse, Button } from 'reactstrap';
import Icon from './icons/Icon';

// https://github.com/HackerNews/API#items
interface HNItem {
  id: number;
  deleted?: boolean;
  type?: 'job' | 'story' | 'comment' | 'poll' | 'pollopt';
  by?: string;
  time?: number; // UNIX time
  text?: string; // HTML
  dead?: boolean; // true if the item is dead.
  parent?: number;
  poll?: number; // a pollopt's poll
  kids?: number[]; // in ranked display order
  url?: string;
  score?: number;
  title?: string; // The title of the story, poll or job. HTML.
  parts?: number[]; // the pollopts of a poll
  descendants?: number;
}

const hNItemLink = (id: number) => `https://news.ycombinator.com/item?id=${id}`;

const buttonBarClasses = "d-inline-flex align-items-center h-border-right py-2 px-2";

const topLevelCommentRefs: RefObject<HTMLElement>[] = [];

const topOfElIsVisible = (el: RefObject<HTMLElement>, buffer = 0) => {
  const top = el.current?.getBoundingClientRect().top;
  if (top === undefined) return false;
  return top < buffer;
}

const LinkToHN = ({ id }: { id: number }) => (
  <a className={`${buttonBarClasses}`} role="button" href={hNItemLink(id)}>
    <Icon name="link" size={1.5} />
  </a>
);

const Share = ({ title, url }: { title?: string, url: string }) => {
  //@ts-ignore
  if (!navigator?.share) return null;

  const shareTo = async () => {
    try {
      //@ts-ignore
      await navigator.share({ title, url });
      console.log('shared!')
    } catch (e) {
      if (e.name === 'AbortError') console.log('Share aborted')
      else console.log(e);
    }
  };

  return (
    <div className={`${buttonBarClasses} clickable`} onClick={shareTo}>
      <Icon name="share" size={1.5} />
    </div>
  );
};

const LinkUrlCard = ({ url }: { url: string }) => (
  <a href={url} className="p-1 mb-2 link-card d-flex align-items-center">
    <Icon size={2} color="gray" name="compass" />
    <span className="pl-2 ml-2 pr-2 link-card--text text-truncate" style={{ flex: '1' }}>{url}</span>
  </a>
);

const Item = ({ id, level = 0 }: { id: number, level?: number }) => {
  const [data, setData] = useState<HNItem>({
    id: id,
    text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    title: 'Lorem ipsum dolor sit amet',
    descendants: 55,
    url: '',
    type: 'comment'
  });

  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    async function getItem() {
      const response = await fetch(`https://hacker-news.firebaseio.com/v0/item/${id}.json`);
      const data = await response.json();

      if (level === 0) document.title = `${data.title} | Dapper`;

      ReactDOM.unstable_batchedUpdates(() => {
        setIsLoading(false);
        /**
         * Sometimes HN returns null for an item, need this so entire page doesn't break
         * TODO error boundry?
         * comment page for: https://news.ycombinator.com/item?id=22359574
         * - https://hacker-news.firebaseio.com/v0/item/22360822.json returns null
         * - https://news.ycombinator.com/item?id=22360822 shows a valid comment
         * (┛ಠ_ಠ)┛彡┻━┻
         */
        setData(data || {
          text: `API error :( <a href="${hNItemLink(id)}">view on hn</a>`,
          type: 'comment'
        });
      })
    }

    getItem();
  }, [id]);

  const containerEl = useRef<HTMLDivElement>(null);
  if (level === 1) topLevelCommentRefs.push(containerEl);

  const [isOpen, setIsOpen] = useState(true);
  const toggle = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    if (level === 0) return;

    // do not collapse comment if clicking on link inside div (HTML from api)
    const target = e.target as HTMLElement;
    if (target.nodeName === 'A') return;

    e.stopPropagation();
    setIsOpen(!isOpen);

    if (containerEl?.current && topOfElIsVisible(containerEl))
      window.scrollTo({ top: containerEl.current.offsetTop - 10, behavior: 'smooth' });
  };

  const levelToColorMap = {
    2: 'blue',
    3: 'gray',
    4: 'purple',
    5: 'green',
    0: 'orange', // use %, so 0 is last (6th)
    1: 'red' // level 1 does not show any, so this is first used on level 7
  }

  // @ts-ignore
  const levelBorderColor = () => `level-border-${levelToColorMap[level % 6]}`;

  const isLoadingClassName = isLoading ? 'loading-skeleton' : '';
  const commentCss = level > 1 ? `${levelBorderColor()} ml-3` : '';
  const topLevel = level === 0;

  return (
    <div ref={containerEl} className={`${commentCss} ${level > 0 ? 'border-top' : ''} px-0 `} onClick={toggle}>
      <div className={`${isLoadingClassName} text-muted small py-1 px-2`}>
        {data.by || '[deleted]'} {/* TODO: Verify what it means if this is empty */}
        {!topLevel && <span className="float-right">
          {isOpen ? String.fromCharCode(8593) : String.fromCharCode(8595)}
        </span>}
      </div>

      <Collapse isOpen={isOpen}>
        <div className={`${isLoadingClassName} py-1 px-2`} style={{ wordBreak: 'break-word' }}>
          {data.deleted && '[deleted]'}
          {data.title && <h4>{data.title}</h4>}
          {data.url && <LinkUrlCard url={data.url} />}
          {data.type === 'poll' && <p>Polls are not supported yet!</p>}
          <div className="item--hn-text" dangerouslySetInnerHTML={{ '__html': DOMPurify.sanitize(data.text || '') }} />
        </div>
        {
          topLevel &&
          <div className="button-bar-background border-top d-flex align-items-center">
            <LinkToHN id={id} />
            <Share title={data.title} url={hNItemLink(id)} />
          </div>
        }
        {data.kids?.map(itemId => <Item id={itemId} key={itemId} level={level + 1} />)}
      </Collapse>
    </div>
  );
}

const ItemPage = ({ id }: { id: number }) => {
  const goToNextComment = () => {
    const firstCommentWithTopVisible = topLevelCommentRefs.filter(e => !topOfElIsVisible(e, 5))[0];

    if (firstCommentWithTopVisible && firstCommentWithTopVisible?.current) {
      window.scrollTo({ top: firstCommentWithTopVisible.current.offsetTop, behavior: 'smooth' });
    }
  }

  return (
    <>
      <Item id={id} />
      <Button size="lg" color="dark" className="rounded-circle floating-button d-md-none" onClick={goToNextComment}>
        &darr;
      </Button>
      <div className="d-md-none" style={{ paddingBottom: '80px' }}></div>
    </>
  );
};

export default ItemPage;
