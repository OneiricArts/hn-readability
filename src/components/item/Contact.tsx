import React from 'react';
import { Link } from 'react-router-dom';
import { LinkHelper } from './LinkHelper';

export const Contact = () => {
  document.title = 'Contact & Appreciation | Dapper';

  return (
    <div className="p-4">
      <Link to="/">&laquo; home</Link>

      <h2>Contact</h2>
      <p>You can email me here:</p>

      <div className="pb-2">
        <img
          src="./email.png"
          alt="oneiric dot arts at gmail dot com"
          width="300px"
        />
      </div>

      <h2>Appreciation</h2>
      <p>
        If you would like to show appreciation, you can send me a{' '}
        <LinkHelper
          openInNewTab
          href="https://www.philzcoffee.com/online-gift-certificate"
          label="Philz Coffee gift card"
        />{' '}
        to the above email. If you cannot afford that, a thank you note is just
        as good.
      </p>

      <p>Alternatively, donate to any of the following charities:</p>
      <ul>
        <li>
          <LinkHelper
            openInNewTab
            href="https://www.unicefusa.org/?form=donate"
            label="UNICEF"
          />
        </li>
      </ul>
    </div>
  );
};
