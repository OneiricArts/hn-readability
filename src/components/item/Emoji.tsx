import React from 'react';

export const Emoji: React.FC<{ emoji: React.ReactNode; label: string }> = ({
  emoji,
  label
}) => (
  <span role="img" aria-label={label}>
    {emoji}
  </span>
);
