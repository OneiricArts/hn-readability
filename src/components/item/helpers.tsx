import { RefObject } from 'react';
import { HNItem } from '../../api/HNApiTypes';
import DOMPurify from 'dompurify';

export const topOfElIsVisible = (el: RefObject<HTMLElement>, buffer = 0) => {
  const top = el.current?.getBoundingClientRect().top;
  if (top === undefined) return false;
  return top < buffer;
};

export const hNItemLink = (id: number) =>
  `https://news.ycombinator.com/item?id=${id}`;

export const setDocumentTitleWithData = (data: HNItem) => {
  if (data.title) {
    document.getElementsByTagName('title')[0].innerHTML = `${DOMPurify.sanitize(
      data.title
    )} | Dapper`;
  } else if (data.text) {
    const div = document.createElement('div');
    div.innerHTML = DOMPurify.sanitize(data.text);
    let title = div.textContent;

    if (title && title?.length > 90) {
      document.title = `${title.substring(0, 90)}... | Dapper`;
    } else {
      document.title = `${title} | Dapper`;
    }
  }
};

/**
 * ONLY FOR DEVELOPMENT USE
 * @param ms time to "sleep" (wait) in milliseconds
 *
 * usage: await sleep()
 */
export const sleep = async (ms: number = 1000) => {
  await new Promise(r => setTimeout(r, ms));
};
