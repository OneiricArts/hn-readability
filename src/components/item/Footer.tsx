import React from 'react';
import { Emoji } from './Emoji';
import { LinkHelper } from './LinkHelper';
import { ResponsiveComponent } from './ResponsiveComponent';

export const Footer = () => (
  <ResponsiveComponent
    xs={
      <div className="h-border-top" style={{ padding: '40px 10px 0px 10px' }}>
        <div className="text-muted small">
          <Content />
        </div>
      </div>
    }
    sm={
      <div className="h-border-top" style={{ padding: '40px 50px 20px' }}>
        <div
          className="text-muted text-center small "
          style={{ paddingTop: '20px' }}
        >
          <Content />
        </div>
      </div>
    }
  />
);

const Content = () => (
  <>
    <img
      src="./logo192.png"
      width="20px"
      alt="Dapper Logo"
      style={{ display: 'inline-block', verticalAlign: 'top' }}
    />
    <span style={{ marginLeft: '-2px' }}>
      apper, a&nbsp;
      <LinkHelper
        openInNewTab
        href="https://news.ycombinator.com"
        label="Hacker News"
      />
      &nbsp;client
    </span>

    <span>&nbsp;|&nbsp;</span>

    <span>
      <LinkHelper
        openInNewTab
        href="https://github.com/OneiricArts/hn-readability"
        label="Github"
      />
    </span>

    <span>&nbsp;|&nbsp;</span>

    <LinkHelper reactRouterLink openInNewTab href="/contact" label="Contact" />

    <span>&nbsp;|&nbsp;</span>

    <LinkHelper
      reactRouterLink
      openInNewTab
      href="/contact"
      label={
        <>
          Buy me a coffee&nbsp;
          <Emoji emoji={<>☕️</>} label="coffee emoji" />
        </>
      }
    />
  </>
);
