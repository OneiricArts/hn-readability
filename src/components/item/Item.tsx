import React, { useEffect, useState, useRef, useContext } from 'react';
import ReactDOM from 'react-dom';
import { HNItem } from '../../api/HNApiTypes';
import { elipsify, hNItemLink } from './helpers';
import getItemFromApi, { IGetItemFromApi } from '../../api/getItemFromApi';
import { ItemCard } from './ItemCard';
import { CommentRefContext } from './ItemPage';
import { Helmet } from 'react-helmet';
import { decode } from 'he';

interface ItemProps {
  id: number;
  level?: number;
  getItem?: IGetItemFromApi;
}

export const Item = ({
  id,
  level = 0,
  getItem = getItemFromApi
}: ItemProps) => {
  const [data, setData] = useState<HNItem>({
    id: id,
    text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    title: 'Lorem ipsum dolor sit amet',
    descendants: 55,
    url: '',
    type: 'comment'
  });

  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    async function getItemAsync() {
      const data = await getItem(id);

      ReactDOM.unstable_batchedUpdates(() => {
        setIsLoading(false);

        setData(
          data || {
            id,
            text: `API error :( <a href="${hNItemLink(id)}">view on hn</a>`,
            error: true
          }
        );
      });
    }

    getItemAsync();
  }, [id, level, getItem]);

  const { addTopLevelCommentRef } = useContext(CommentRefContext);
  const containerEl = useRef<HTMLDivElement>(null);
  if (level === 1) addTopLevelCommentRef(containerEl);

  if (level === 0) console.log(data.title);

  return (
    <>
      {/* TODO remove !isLoading when placeholder data no longer needed */}
      {!isLoading && level === 0 && (
        <Helmet titleTemplate="%s | Dapper">
          <title>
            {decode(data.title || (data.text && elipsify(data.text, 90)) || '')}
          </title>
        </Helmet>
      )}

      <ItemCard
        containerEl={containerEl}
        data={data}
        level={level}
        isLoading={isLoading}
        kids={data.kids?.map(itemId => (
          <Item id={itemId} key={itemId} level={level + 1} />
        ))}
      />
    </>
  );
};
