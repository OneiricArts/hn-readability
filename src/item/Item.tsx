import React, { useEffect, useState, useRef, RefObject } from 'react';
import ReactDOM from 'react-dom';
import DOMPurify from 'dompurify';
import { Collapse } from 'reactstrap';
import Icon from '../icons/Icon';
import { Link } from 'react-router-dom';
import Parent from './Parent';
import { HNItem } from '../HNApiTypes';
import { topOfElIsVisible } from './helpers';
import { Share } from './Share';
import { TimeAgo } from '../timeago';
import LinkUrlCard from './LinkUrlCard';
import LinksToHn from './linksToHn';

const hNItemLink = (id: number) => `https://news.ycombinator.com/item?id=${id}`;

export const buttonBarClasses =
  'd-inline-flex align-items-center h-border-right py-2 px-2';

const LinkToHN = ({ id }: { id: number }) => (
  <a
    className={`${buttonBarClasses} hnr-inherit-color`}
    role="button"
    href={hNItemLink(id)}
  >
    <Icon name="link" size={1.5} />
  </a>
);

interface ItemProps {
  id: number;
  level?: number;
  addTopLevelCommentRef: (r: RefObject<HTMLElement>) => void;
}

export const Item = ({ id, level = 0, addTopLevelCommentRef }: ItemProps) => {
  const [data, setData] = useState<HNItem>({
    id: id,
    text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    title: 'Lorem ipsum dolor sit amet',
    descendants: 55,
    url: '',
    type: 'comment'
  });

  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    async function getItem() {
      const response = await fetch(
        `https://hacker-news.firebaseio.com/v0/item/${id}.json`
      );
      const data: HNItem = await response.json();

      if (level === 0) {
        if (data.title) {
          document.getElementsByTagName(
            'title'
          )[0].innerHTML = `${DOMPurify.sanitize(data.title)} | Dapper`;
        } else if (data.text) {
          const div = document.createElement('div');
          div.innerHTML = DOMPurify.sanitize(data.text);
          let title = div.textContent;

          if (title && title?.length > 90) {
            document.title = `${title.substring(0, 90)}... | Dapper`;
          } else {
            document.title = `${title} | Dapper`;
          }
        }
      }

      ReactDOM.unstable_batchedUpdates(() => {
        setIsLoading(false);
        /**
         * Sometimes HN returns null for an item, need this so entire page doesn't break
         * TODO error boundry?
         * comment page for: https://news.ycombinator.com/item?id=22359574
         * - https://hacker-news.firebaseio.com/v0/item/22360822.json returns null
         * - https://news.ycombinator.com/item?id=22360822 shows a valid comment
         * (┛ಠ_ಠ)┛彡┻━┻
         */
        setData(
          data || {
            text: `API error :( <a href="${hNItemLink(id)}">view on hn</a>`,
            type: 'comment'
          }
        );
      });
    }

    getItem();
  }, [id, level]);

  const containerEl = useRef<HTMLDivElement>(null);
  if (level === 1) addTopLevelCommentRef(containerEl);

  const [isOpen, setIsOpen] = useState(true);
  const toggle = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    if (level === 0) return;

    // do not collapse comment if user is selecting text
    const selection = window.getSelection();
    if (selection?.type === 'Range') return;

    // do not collapse comment if clicking on link inside div (HTML from api), or link icon, or dapper link at bottom
    const target = e.target as HTMLElement;
    if (target.nodeName === 'A' || target.nodeName === 'svg') return;
    if (
      target.parentNode?.nodeName === 'A' ||
      target.parentNode?.nodeName === 'svg'
    )
      return;

    e.stopPropagation();
    setIsOpen(!isOpen);

    if (containerEl?.current && topOfElIsVisible(containerEl))
      window.scrollTo({
        top: containerEl.current.offsetTop - 10,
        behavior: 'smooth'
      });
  };

  const levelToColorMap = {
    2: 'blue',
    3: 'gray',
    4: 'purple',
    5: 'green',
    0: 'orange', // use %, so 0 is last (6th)
    1: 'red' // level 1 does not show any, so this is first used on level 7
  };

  // @ts-ignore
  const levelBorderColor = () => `level-border-${levelToColorMap[level % 6]}`;

  const isLoadingClassName = isLoading ? 'loading-skeleton' : '';
  const commentCss = level > 1 ? `${levelBorderColor()} ml-3` : '';
  const topLevel = level === 0;

  return (
    <div
      ref={containerEl}
      className={`${commentCss} ${level > 0 ? 'h-border-top' : ''} px-0 `}
      onClick={toggle}
    >
      <div
        className={`${isLoadingClassName} text-muted small pt-1 px-2`}
        style={{ display: 'flex', alignItems: 'center' }}
      >
        {data.by || '[deleted]'}{' '}
        {/* TODO: Verify what it means if this is empty */}
        {topLevel && (
          <>
            &emsp;{data.score && `↑${data.score}`}&emsp;
            <TimeAgo time={data.time} icon />
          </>
        )}
        {!topLevel && (
          <span className="ml-auto">
            <Link to={`/item?id=${data.id}`} className="mr-3 hnr-inherit-color">
              <Icon name="link" svgClassName="ignore" />
            </Link>
            <TimeAgo time={data.time} />
            &nbsp;
            {isOpen ? String.fromCharCode(8593) : String.fromCharCode(8595)}
          </span>
        )}
      </div>

      {topLevel && (
        <div className="px-2 py-1 text-muted small w-100 text-truncate">
          {data.parent && <Parent parent={data.parent} />}
        </div>
      )}

      <Collapse isOpen={isOpen}>
        <div
          className={`${isLoadingClassName} py-1 px-2`}
          style={{ wordBreak: 'break-word' }}
        >
          {data.deleted && '[deleted]'}
          {data.title && (
            <h4
              dangerouslySetInnerHTML={{
                __html: DOMPurify.sanitize(data.title)
              }}
            />
          )}
          {data.url && <LinkUrlCard url={data.url} />}
          {data.type === 'poll' && <p>Polls are not supported yet!</p>}
          <div
            className="item--hn-text"
            dangerouslySetInnerHTML={{
              __html: DOMPurify.sanitize(data.text || '')
            }}
          />
        </div>

        {/* If item contains link to hacker news item,
        append link cards to dapper url of item */}
        {data.text && <LinksToHn text={data.text} />}

        {topLevel && (
          <div className="h-border-top d-flex align-items-center">
            <LinkToHN id={id} />
            <Share title={data.title} url={hNItemLink(id)} />
          </div>
        )}
        {data.kids?.map(itemId => (
          <Item
            id={itemId}
            key={itemId}
            level={level + 1}
            addTopLevelCommentRef={addTopLevelCommentRef}
          />
        ))}
      </Collapse>
    </div>
  );
};
