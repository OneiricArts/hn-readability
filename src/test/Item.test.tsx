import React from 'react';
import { Item } from '../item/Item';
import { render, act, RenderResult } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import sinon from 'sinon';
import mockdate from 'mockdate';

import item8863 from './resources/8863';
import item23326827 from './resources/23326827';
import item23325786 from './resources/23325786';

/**
 * TODO
 *
 * Commented out all kids for mocked items
 * so I don't have to mock all recursive calls.
 *
 * Instead, should snapshot test a UI component after refactors.
 */

/**
 * @jest-environment jsdom
 */

describe('Item', () => {
  beforeEach(() => {
    mockdate.set('6/1/2020');

    // @ts-ignore
    global.document.getElementsByTagName = jest
      .fn()
      .mockReturnValue([{ innerHTML: {} }]);
  });

  afterEach(() => {
    mockdate.reset();
  });

  it('snapshot tests', async () => {
    const testCases = [
      { id: 8863, response: item8863 },
      { id: 23326827, response: item23326827 },
      { id: 23325786, response: item23325786 }
    ];

    for (let i = 0; i < testCases.length; i += 1) {
      const id = testCases[i].id;
      const response = testCases[i].response;

      //@ts-ignore
      let wrapper: RenderResult = undefined;

      await act(async () => {
        wrapper = render(
          <BrowserRouter>
            <Item
              id={id}
              addTopLevelCommentRef={() => undefined}
              getItem={sinon.stub().withArgs(id).resolves(response)}
            />
          </BrowserRouter>
        );
      });

      expect(wrapper.container.firstChild).toMatchSnapshot();
    }
  });
});
