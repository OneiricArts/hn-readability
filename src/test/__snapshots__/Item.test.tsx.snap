// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item snapshot tests 1`] = `
<div
  class="  px-0 "
>
  <div
    class=" text-muted small pt-1 px-2"
    style="display: flex; align-items: center;"
  >
    dhouston
     
     
    ↑104
     
    <svg
      height="1.1em"
      style="display: inline-block; vertical-align: middle;"
      viewBox="0 0 512 512"
      width="1.1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 64C150 64 64 150 64 256s86 192 192 192 192-86 192-192S362 64 256 64z"
        fill="none"
        stroke="currentColor"
        stroke-miterlimit="10"
        stroke-width="32"
      />
      <path
        d="M256 128L256 272 352 272"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="32"
      />
    </svg>
    13.2y
  </div>
  <div
    class="px-2 py-1 text-muted small w-100 text-truncate"
  />
  <div
    class="collapse show"
  >
    <div
      class=" py-1 px-2"
      style="word-break: break-word;"
    >
      <h4>
        My YC app: Dropbox - Throw away your USB drive
      </h4>
      <a
        class="p-1 link-card d-flex align-items-center mb-2"
        href="http://www.getdropbox.com/u/2/screencast.html"
      >
        <svg
          height="2em"
          style="display: inline-block; vertical-align: middle;"
          viewBox="0 0 512 512"
          width="2em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M448 256c0-106-86-192-192-192S64 150 64 256s86 192 192 192 192-86 192-192z"
            fill="none"
            stroke="currentColor"
            stroke-miterlimit="10"
            stroke-width="32"
          />
          <path
            d="M350.67 150.93l-117.2 46.88a64 64 0 00-35.66 35.66l-46.88 117.2a8 8 0 0010.4 10.4l117.2-46.88a64 64 0 0035.66-35.66l46.88-117.2a8 8 0 00-10.4-10.4zM256 280a24 24 0 1124-24 24 24 0 01-24 24z"
            fill="currentColor"
          />
        </svg>
        <span
          class="pl-2 ml-2 pr-2 link-card--text text-truncate"
          style="flex: 1;"
        >
          http://www.getdropbox.com/u/2/screencast.html
        </span>
      </a>
      <div
        class="item--hn-text"
      />
    </div>
    <div
      class="h-border-top d-flex align-items-center"
    >
      <a
        class="d-inline-flex align-items-center h-border-right py-2 px-2 hnr-inherit-color"
        href="https://news.ycombinator.com/item?id=8863"
        role="button"
      >
        <svg
          height="1.5em"
          style="display: inline-block; vertical-align: middle;"
          viewBox="0 0 512 512"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M208 352h-64a96 96 0 010-192h64M304 160h64a96 96 0 010 192h-64"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="36"
          />
          <path
            d="M163.29 256L350.71 256"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="36"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;

exports[`Item snapshot tests 2`] = `
<div
  class="  px-0 "
>
  <div
    class=" text-muted small pt-1 px-2"
    style="display: flex; align-items: center;"
  >
    Diederich
     
     
     
    <svg
      height="1.1em"
      style="display: inline-block; vertical-align: middle;"
      viewBox="0 0 512 512"
      width="1.1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 64C150 64 64 150 64 256s86 192 192 192 192-86 192-192S362 64 256 64z"
        fill="none"
        stroke="currentColor"
        stroke-miterlimit="10"
        stroke-width="32"
      />
      <path
        d="M256 128L256 272 352 272"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="32"
      />
    </svg>
    4d
  </div>
  <div
    class="px-2 py-1 text-muted small w-100 text-truncate"
  />
  <div
    class="collapse show"
  >
    <div
      class=" py-1 px-2"
      style="word-break: break-word;"
    >
      <div
        class="item--hn-text"
      >
        Yup; it's packed full of such gems.
      </div>
    </div>
    <div
      class="h-border-top d-flex align-items-center"
    >
      <a
        class="d-inline-flex align-items-center h-border-right py-2 px-2 hnr-inherit-color"
        href="https://news.ycombinator.com/item?id=23326827"
        role="button"
      >
        <svg
          height="1.5em"
          style="display: inline-block; vertical-align: middle;"
          viewBox="0 0 512 512"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M208 352h-64a96 96 0 010-192h64M304 160h64a96 96 0 010 192h-64"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="36"
          />
          <path
            d="M163.29 256L350.71 256"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="36"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;

exports[`Item snapshot tests 3`] = `
<div
  class="  px-0 "
>
  <div
    class=" text-muted small pt-1 px-2"
    style="display: flex; align-items: center;"
  >
    r721
     
     
     
    <svg
      height="1.1em"
      style="display: inline-block; vertical-align: middle;"
      viewBox="0 0 512 512"
      width="1.1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 64C150 64 64 150 64 256s86 192 192 192 192-86 192-192S362 64 256 64z"
        fill="none"
        stroke="currentColor"
        stroke-miterlimit="10"
        stroke-width="32"
      />
      <path
        d="M256 128L256 272 352 272"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="32"
      />
    </svg>
    4d
  </div>
  <div
    class="px-2 py-1 text-muted small w-100 text-truncate"
  />
  <div
    class="collapse show"
  >
    <div
      class=" py-1 px-2"
      style="word-break: break-word;"
    >
      <div
        class="item--hn-text"
      >
        Recent discussions:
        <p>
          <a
            href="https://news.ycombinator.com/item?id=23317570"
            rel="nofollow"
          >
            https://news.ycombinator.com/item?id=23317570
          </a>
        </p>
        <p>
          <a
            href="https://news.ycombinator.com/item?id=23223219"
            rel="nofollow"
          >
            https://news.ycombinator.com/item?id=23223219
          </a>
        </p>
        <p>
          <a
            href="https://news.ycombinator.com/item?id=23221264"
            rel="nofollow"
          >
            https://news.ycombinator.com/item?id=23221264
          </a>
        </p>
      </div>
    </div>
    <div
      class="py-2 mx-2"
    >
      <a
        class="p-1 link-card d-flex align-items-center "
        href="https://dapper.dilraj.dev/item?id=23317570"
      >
        <svg
          height="2em"
          style="display: inline-block; vertical-align: middle;"
          viewBox="0 0 512 512"
          width="2em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M448 256c0-106-86-192-192-192S64 150 64 256s86 192 192 192 192-86 192-192z"
            fill="none"
            stroke="currentColor"
            stroke-miterlimit="10"
            stroke-width="32"
          />
          <path
            d="M350.67 150.93l-117.2 46.88a64 64 0 00-35.66 35.66l-46.88 117.2a8 8 0 0010.4 10.4l117.2-46.88a64 64 0 0035.66-35.66l46.88-117.2a8 8 0 00-10.4-10.4zM256 280a24 24 0 1124-24 24 24 0 01-24 24z"
            fill="currentColor"
          />
        </svg>
        <span
          class="pl-2 ml-2 pr-2 link-card--text text-truncate"
          style="flex: 1;"
        >
          https://dapper.dilraj.dev/item?id=23317570
        </span>
      </a>
    </div>
    <div
      class="py-2 mx-2"
    >
      <a
        class="p-1 link-card d-flex align-items-center "
        href="https://dapper.dilraj.dev/item?id=23223219"
      >
        <svg
          height="2em"
          style="display: inline-block; vertical-align: middle;"
          viewBox="0 0 512 512"
          width="2em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M448 256c0-106-86-192-192-192S64 150 64 256s86 192 192 192 192-86 192-192z"
            fill="none"
            stroke="currentColor"
            stroke-miterlimit="10"
            stroke-width="32"
          />
          <path
            d="M350.67 150.93l-117.2 46.88a64 64 0 00-35.66 35.66l-46.88 117.2a8 8 0 0010.4 10.4l117.2-46.88a64 64 0 0035.66-35.66l46.88-117.2a8 8 0 00-10.4-10.4zM256 280a24 24 0 1124-24 24 24 0 01-24 24z"
            fill="currentColor"
          />
        </svg>
        <span
          class="pl-2 ml-2 pr-2 link-card--text text-truncate"
          style="flex: 1;"
        >
          https://dapper.dilraj.dev/item?id=23223219
        </span>
      </a>
    </div>
    <div
      class="py-2 mx-2"
    >
      <a
        class="p-1 link-card d-flex align-items-center "
        href="https://dapper.dilraj.dev/item?id=23221264"
      >
        <svg
          height="2em"
          style="display: inline-block; vertical-align: middle;"
          viewBox="0 0 512 512"
          width="2em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M448 256c0-106-86-192-192-192S64 150 64 256s86 192 192 192 192-86 192-192z"
            fill="none"
            stroke="currentColor"
            stroke-miterlimit="10"
            stroke-width="32"
          />
          <path
            d="M350.67 150.93l-117.2 46.88a64 64 0 00-35.66 35.66l-46.88 117.2a8 8 0 0010.4 10.4l117.2-46.88a64 64 0 0035.66-35.66l46.88-117.2a8 8 0 00-10.4-10.4zM256 280a24 24 0 1124-24 24 24 0 01-24 24z"
            fill="currentColor"
          />
        </svg>
        <span
          class="pl-2 ml-2 pr-2 link-card--text text-truncate"
          style="flex: 1;"
        >
          https://dapper.dilraj.dev/item?id=23221264
        </span>
      </a>
    </div>
    <div
      class="h-border-top d-flex align-items-center"
    >
      <a
        class="d-inline-flex align-items-center h-border-right py-2 px-2 hnr-inherit-color"
        href="https://news.ycombinator.com/item?id=23325786"
        role="button"
      >
        <svg
          height="1.5em"
          style="display: inline-block; vertical-align: middle;"
          viewBox="0 0 512 512"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M208 352h-64a96 96 0 010-192h64M304 160h64a96 96 0 010 192h-64"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="36"
          />
          <path
            d="M163.29 256L350.71 256"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="36"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;
